types/uint.go

```diff
@@ -0,0 +1,177 @@
+func (u Uint) Add(u2 Uint) Uint { return NewUintFromBigInt(new(big.Int).Add(u.i, u2.i)) }
+
+// Add convert uint64 and add it to Uint
+func (u Uint) AddUint64(u2 uint64) Uint { return u.Add(NewUint(u2)) }
+
+// Add adds Uint from another
+func (u Uint) Sub(u2 Uint) Uint { return NewUintFromBigInt(new(big.Int).Sub(u.i, u2.i)) }
+
+// Add adds Uint from another
+func (u Uint) SubUint64(u2 uint64) Uint { return u.Sub(NewUint(u2)) }
+
+// Mul multiples two Uints
+func (u Uint) Mul(u2 Uint) (res Uint) {
+	return NewUintFromBigInt(new(big.Int).Mul(u.i, u2.i))
+}
+
+// Mul multiples two Uints
+func (u Uint) MulUint64(u2 uint64) (res Uint) { return u.Mul(NewUint(u2)) }
+
+// Div divides Uint with Uint
+func (u Uint) Div(u2 Uint) (res Uint) { return NewUintFromBigInt(div(u.i, u2.i)) }
+
+// Div divides Uint with uint64
+func (u Uint) DivUint64(u2 uint64) Uint { return u.Div(NewUint(u2)) }
+
+// Return the minimum of the Uints
+func MinUint(u1, u2 Uint) Uint { return NewUintFromBigInt(min(u1.i, u2.i)) }
+
+// Return the maximum of the Uints
+func MaxUint(u1, u2 Uint) Uint { return NewUintFromBigInt(max(u1.i, u2.i)) }
+
+// Human readable string
+func (u Uint) String() string { return u.i.String() }
+
+// Testing purpose random Uint generator
+func randomUint(u Uint) Uint { return NewUintFromBigInt(random(u.i)) }
+
+// MarshalAmino defines custom encoding scheme
+func (u Uint) MarshalAmino() (string, error) {
+	if u.i == nil { // Necessary since default Uint initialization has i.i as nil
+		u.i = new(big.Int)
+	}
+	return marshalAmino(u.i)
+}
+
+// UnmarshalAmino defines custom decoding scheme
+func (u *Uint) UnmarshalAmino(text string) error {
+	if u.i == nil { // Necessary since default Uint initialization has i.i as nil
+		u.i = new(big.Int)
+	}
+	return unmarshalAmino(u.i, text)
+}
+
+// MarshalJSON defines custom encoding scheme
+func (u Uint) MarshalJSON() ([]byte, error) {
+	if u.i == nil { // Necessary since default Uint initialization has i.i as nil
+		u.i = new(big.Int)
+	}
+	return marshalJSON(u.i)
+}
+
+// UnmarshalJSON defines custom decoding scheme
+func (u *Uint) UnmarshalJSON(bz []byte) error {
+	if u.i == nil { // Necessary since default Uint initialization has i.i as nil
+		u.i = new(big.Int)
+	}
+	return unmarshalJSON(u.i, bz)
+}
+
+//__________________________________________________________________________
+
+// UintOverflow returns true if a given unsigned integer overflows and false
+// otherwise.
+func UintOverflow(i *big.Int) error {
+	if i.Sign() < 0 {
+		return errors.New("non-positive integer")
+	}
+	if i.BitLen() > 256 {
+		return fmt.Errorf("bit length %d greater than 256", i.BitLen())
+	}
+	return nil
+}
+
+// ParseUint reads a string-encoded Uint value and return a Uint.
+func ParseUint(s string) (Uint, error) {
+	i, ok := new(big.Int).SetString(s, 0)
+	if !ok {
+		return Uint{}, fmt.Errorf("cannot convert %q to big.Int", s)
+	}
+	return checkNewUint(i)
+}
+
+func checkNewUint(i *big.Int) (Uint, error) {
```

Comment:
It seems this function is somewhat redundant. In the constructor(s) we should just check `UintOverflow` since we panic anyway on error. In `ParseUint`, we should also just check `UintOverflow`.

Reply:
`checkNewUint` is there to avoid calling `UintOverflow()` twice when constructing a new `Uint` through `ParseUint`. `checkNewUint` builds a new `Uint` and returns `UintOverflow` result. Whilst `ParseUint` reads the result and returns it the caller, `New` just panics if the result is not nil.

Reply:
Ok, fair enough. Disregard.